test/TransFNeg.ll

```diff
@@ -0,0 +1,41 @@
+; RUN: llvm-as %s -o %t.bc
```

Comment:
For LLVM 8.0.0 and onwards, this should be a transcoding test, and `test/OpFNegate.spvasm` should be removed. Recent releases of Clang generate **fneg** for CL source, so this should be implemented using OpenCL C as input, unless there is a good reason not to: I would recommend `DivRem.cl` as an example.

(If this test _does_ have to continue to use LLVM IR as input for any reason, then the IR should be simplified, e.g. there should be no need to use **alloca**. If needed, it is possible that compiling `neg.cl` with `-O1` or `-O2` might provide a better starting point.)

The test should cover a few different types. For example, we should have one test with a single precision float, another with a double precision float, and a third with a vector of half precision floats. (I admit that there is an aspect of "do as I say, not as I do" to this: the SPIR-V assembly tests that I added only covered scalars, with the exception of the `OpFMod` and `OpSMod` tests that were added later. However, any new tests that are added should cover both, unless there is a strong reason not to.)

If LLVM's **fneg** instruction guarantees that **fneg**(-0.0) = +0.0 and **fneg**(+0.0) = -0.0, then this test otherwise looks okay in principle. It is not clear to me whether this change alters the handling of `OpFNegate` in `SPIRVReader`. Could you confirm this, please?

We should ensure we use `UnaryOperator::CreateFNeg()` instead of `BinaryOperator::CreateFNeg()`, either now, or in future versions of the translator. As noted above, recent versions of Clang do generate **fneg**, although there is a comment within LLVM which suggests that the use of LLVM's **fneg** instruction as output may result in optimizations being missed. It is not clear whether this comment is still accurate, but I would tend to assume not, if Clang now generates the instruction.

Note that LLVM 7.0.x does not provide an **fneg** instruction. It would be helpful to mention that it was introduced in LLVM 8.0.0 in the commit message and in the pull request comment. I believe that the logic of subtracting from -0.0 when using LLVM 7.0.x matches the semantics of `OpFNegate`, i.e. for any floating point value _f_, **signbit**(**fneg**(_f_)) ≠ **fneg**(_f_). Can you confirm this, please?

Reply:
> Thanks for the suggestion, I've generated reproducer for #359 from OpenCL C using `clang-10`, but I had to manually edit the IR to get `fneg` there.

Are you using a recent version of the Clang development tree? The version I'm using generates **fneg** with no manual modifications needed to the IR. If I use Clang 9, then I find that an **fsub** is generated instead.

> Looking at [lang ref](https://llvm.org/docs/LangRef.html#fneg-instruction) I cannot find exact comments about this, it says the following:
> 
> > The value produced is a copy of the operand with its sign bit flipped.
> 
> So, I guess fneg(-0.0) = +0.0 is true with fneg.

Yes, I came to the same conclusion. It seems intended to correspond to the **negate** operation in IEEE 754-2008/2019 § 5.5.1 _Sign bit operations_, which is defined as follows:

> "**negate**(_x_) copies a floating-point operand _x_ to a destination in the same format, reversing the sign bit. **negate**(_x_) is not the same as **subtraction**(0, _x_)

My understanding is that **negate**(_x_) _is_ the same as **subtraction**(−0, _x_) for rounding modes except **roundTowardNegative**.

> According to SPIR-V spec, `OpFNegate`:
> 
> > Floating-point subtract of Operand from zero.

There have been cases in the past where a SPIR-V operation was intended to align with an LLVM instruction and/or IEEE 754 semantics, and we have sought clarification. It is worth considering whether `OpFNegate` is really intended to be different to the LLVM **fneg** operation (which is the same as the IEEE 754-2008/2019 **negate** operation).

> So, I think that our current implementation with fsub is correct. We might want to update it to use fneg instead but I'm not sure that this is something critical that should be resolved immediately.

If `OpFNegate` and **fneg** are intended to be the same, then it would be appropriate to translate `OpFNegate` to LLVM's **fneg** operation, and to translate LLVM's **fneg** operation to `OpFNegate`. However, it would not be appropriate to translate `OpFNegate` to **fsub** for the **roundTowardNegative** rounding mode.

If `OpFNegate` and **fneg** are intended to be different, then it would not be appropriate to translate **fneg** (with precise IEEE 754-2008 sematics) into `OpFNegate` (with possibly looser semantics).

> Would it be okay with you if we submit an issue to this repo to consider generating `fneg` from `OpFNegate`?

I'm absolutely okay with this being done in stages. It seems completely reasonable to me to add support for translation of **fneg** to `OpFNegate`, as a first step.

If the amount of extra work needed is minimal (as I believe it is) then I think it may be preferable to make all of the necessary changes in this pull request, assuming that the question of the exact semantics of `OpFNegate` is not a barrier.

However, if **fneg** is poorly optimized for older versions of LLVM, and it would simplify backporting to the {{llvm_release_90}} and/or {{llvm_release_80}} branches to separate out any changes relating to translation from SPIR-V to LLVM IR (allowing the second part to be dropped when backporting), then these changes *should* be made separately.

> According to SPIR-V spec, `OpFNegate`:
> 
> > Floating-point subtract of Operand from zero.
> 
> If I understand correctly, this is exact match to LLVM 7.0.x IR

LLVM's `BinaryOperator::CreateFNeg()` creates a subtraction from _negative_ zero, which is correct for all rounding modes except **roundTowardNegative**. Although it would be possible to construct something that works for **roundTowardNegative** with LLVM 7.0.x, it would not be worthwhile, especially as it would result in poorer performing code, and so I think the existing **fsub** would have to suffice.